File: python/paylabs_signature.py
import hashlib
import base64
import json
import os
from datetime import datetime, timezone, timedelta
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding


def remove_nulls(obj):
    """Recursively remove null values from dict/list."""
    if isinstance(obj, dict):
        return {k: remove_nulls(v) for k, v in obj.items() if v is not None}
    elif isinstance(obj, list):
        return [remove_nulls(item) for item in obj]
    return obj


def minify_json(body):
    """Minify JSON without whitespace."""
    cleaned = remove_nulls(body)
    return json.dumps(cleaned, separators=(',', ':'))


def sha256_hex(data):
    """Generate SHA256 hash in lowercase hex."""
    return hashlib.sha256(data.encode()).hexdigest().lower()


def generate_timestamp():
    """Generate timestamp in WIB timezone (UTC+7)."""
    wib = timezone(timedelta(hours=7))
    now = datetime.now(wib)
    return now.strftime('%Y-%m-%dT%H:%M:%S') + '+07:00'


def generate_request_id():
    """Generate unique request ID with timestamp and random number."""
    import random
    wib = timezone(timedelta(hours=7))
    now = datetime.now(wib)
    date_str = now.strftime('%Y%m%d%H%M%S')
    random_str = str(random.randint(0, 999999)).zfill(6)
    return f"{date_str}{random_str}"


def parse_private_key(pem_str):
    """Parse PEM private key string."""
    pem_str = pem_str.replace('\\n', '\n')
    return serialization.load_pem_private_key(pem_str.encode(), password=None)


def parse_public_key(pem_str):
    """Parse PEM public key string."""
    pem_str = pem_str.replace('\\n', '\n')
    return serialization.load_pem_public_key(pem_str.encode())


def generate_signature(method, endpoint, body, private_key_pem):
    """Generate RSA-SHA256 signature for Paylabs API."""
    timestamp = generate_timestamp()
    
    body_hash = sha256_hex(minify_json(body))
    print(minify_json(body))
    
    string_to_sign = f"{method}:{endpoint}:{body_hash}:{timestamp}"
    print(f"String to Sign: {string_to_sign}")
    
    private_key = parse_private_key(private_key_pem)
    
    signature_bytes = private_key.sign(
        string_to_sign.encode(),
        padding.PKCS1v15(),
        hashes.SHA256()
    )
    
    signature = base64.b64encode(signature_bytes).decode()
    print(f"Generated Signature: {signature}")
    
    return signature, timestamp


def verify_signature(string_to_verify, signature_base64, public_key_pem):
    """Verify RSA-SHA256 signature from Paylabs callback."""
    try:
        public_key = parse_public_key(public_key_pem)
        signature = base64.b64decode(signature_base64)
        
        public_key.verify(
            signature,
            string_to_verify.encode(),
            padding.PKCS1v15(),
            hashes.SHA256()
        )
        return True
    except Exception as e:
        print(f"Verification failed: {e}")
        return False


def log_to_local_server(endpoint, request_headers, request_body, response_body):
    """Log outbound request/response to the visualizer local server."""
    import requests
    try:
        log_data = {
            'type': 'outbound',
            'endpoint': endpoint,
            'requestHeaders': request_headers,
            'requestBody': request_body,
            'responseBody': response_body
        }
        log_port = os.getenv('PORT', '3000')
        requests.post(f'http://localhost:{log_port}/log', json=log_data, timeout=2)
    except Exception:
        # Silently fail if visualizer is not running
        pass


def create_transaction(endpoint, body):
    """Generic function to create a transaction."""
    import requests
    from dotenv import load_dotenv
    
    load_dotenv()
    
    base_url = os.getenv('PAYLABS_BASE_URL')
    merchant_id = os.getenv('MERCHANT_ID')
    private_key = os.getenv('MERCHANT_PRIVATE_KEY')
    
    signature, timestamp = generate_signature('POST', endpoint, body, private_key)
    
    headers = {
        'Content-Type': 'application/json',
        'X-PARTNER-ID': merchant_id,
        'X-TIMESTAMP': timestamp,
        'X-SIGNATURE': signature,
        'X-REQUEST-ID': body.get('requestId'),
    }
    
    try:
        minified_body = minify_json(body)
        response = requests.post(
            f'{base_url}{endpoint}',
            data=minified_body,
            headers=headers
        )
        print(json.dumps(response.json(), indent=2))
        return response.json()
    except Exception as e:
        print(f'Error: {e}')


def get_public_ip():
    """Get public IP address."""
    import requests
    try:
        response = requests.get('https://api.ipify.org?format=json', timeout=5)
        return response.json().get('ip', '127.0.0.1')
    except Exception:
        return '127.0.0.1'


def create_transaction_snap(endpoint, body):
    """Create SNAP Transaction."""
    import requests
    from dotenv import load_dotenv
    
    load_dotenv()
    
    base_url = os.getenv('PAYLABS_BASE_URL')
    merchant_id = os.getenv('MERCHANT_ID')
    private_key = os.getenv('MERCHANT_PRIVATE_KEY')
    
    ip_address = get_public_ip()
    print(f"Public IP: {ip_address}")
    
    # SNAP signature uses a modified endpoint logic if needed
    signature_endpoint = endpoint
    if endpoint.startswith("/api/v1.0"):
        signature_endpoint = endpoint.replace("/api/v1.0", "")
        
    # Remove requestId/externalId from body before signing/sending for SNAP
    external_id = body.pop('externalId', None)
    if not external_id:
        external_id = body.pop('requestId', None)
    
    if not external_id:
        external_id = generate_request_id()
        
    signature, timestamp = generate_signature('POST', signature_endpoint, body, private_key)
    
    headers = {
        'Content-Type': 'application/json;charset=utf-8',
        'X-PARTNER-ID': merchant_id,
        'X-TIMESTAMP': timestamp,
        'X-SIGNATURE': signature,
        'X-EXTERNAL-ID': external_id,
        'X-IP-ADDRESS': ip_address,
    }
    
    try:
        minified_body = minify_json(body)
        response = requests.post(
            f'{base_url}{endpoint}',
            data=minified_body,
            headers=headers
        )
        print(json.dumps(response.json(), indent=2))
        return response.json()
    except Exception as e:
        print(f'Error: {e}')

File: php/PaylabsSignature.php
<?php

class PaylabsSignature
{
    /**
     * Remove null values from array recursively.
     */
    public static function removeNulls($data)
    {
        if (is_array($data)) {
            $result = [];
            foreach ($data as $key => $value) {
                if ($value !== null) {
                    $result[$key] = self::removeNulls($value);
                }
            }
            return $result;
        }
        return $data;
    }

    /**
     * Minify JSON without whitespace.
     */
    public static function minifyJson($body)
    {
        $cleaned = self::removeNulls($body);
        return json_encode($cleaned, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
    }

    /**
     * Generate SHA256 hash in lowercase hex.
     */
    public static function sha256Hex($data)
    {
        return strtolower(hash('sha256', $data));
    }

    /**
     * Generate timestamp in WIB timezone (UTC+7).
     */
    public static function generateTimestamp()
    {
        $tz = new DateTimeZone('Asia/Jakarta');
        $now = new DateTime('now', $tz);
        return $now->format('Y-m-d\TH:i:s') . '+07:00';
    }

    /**
     * Generate unique request ID with timestamp and random number.
     */
    public static function generateRequestId()
    {
        $tz = new DateTimeZone('Asia/Jakarta');
        $now = new DateTime('now', $tz);
        $dateStr = $now->format('YmdHis');
        $randomStr = str_pad(rand(0, 999999), 6, '0', STR_PAD_LEFT);
        return $dateStr . $randomStr;
    }

    /**
     * Generate RSA-SHA256 signature for Paylabs API.
     */
    public static function generateSignature($method, $endpoint, $body, $privateKeyPem)
    {
        $timestamp = self::generateTimestamp();
        
        $bodyHash = self::sha256Hex(self::minifyJson($body));
        echo self::minifyJson($body) . "\n";
        
        $stringToSign = "{$method}:{$endpoint}:{$bodyHash}:{$timestamp}";
        echo "String to Sign: {$stringToSign}\n";
        
        $privateKeyPem = str_replace('\\n', "\n", $privateKeyPem);
        $privateKey = openssl_pkey_get_private($privateKeyPem);
        
        openssl_sign($stringToSign, $signature, $privateKey, OPENSSL_ALGO_SHA256);
        
        $signatureBase64 = base64_encode($signature);
        echo "Generated Signature: {$signatureBase64}\n";
        
        return [$signatureBase64, $timestamp];
    }

    /**
     * Verify RSA-SHA256 signature from Paylabs callback.
     */
    public static function verifySignature($stringToVerify, $signatureBase64, $publicKeyPem)
    {
        try {
            $publicKeyPem = str_replace('\\n', "\n", $publicKeyPem);
            $publicKey = openssl_pkey_get_public($publicKeyPem);
            
            $signature = base64_decode($signatureBase64);
            
            $result = openssl_verify($stringToVerify, $signature, $publicKey, OPENSSL_ALGO_SHA256);
            
            return $result === 1;
        } catch (Exception $e) {
            echo "Verification failed: " . $e->getMessage() . "\n";
            return false;
        }
    }

    /**
     * Log outbound request/response to the visualizer local server.
     */
    public static function logToLocalServer($endpoint, $requestHeaders, $requestBody, $responseBody)
    {
        try {
            $logPort = $_ENV['PORT'] ?? '3000';
            $logData = [
                'type' => 'outbound',
                'endpoint' => $endpoint,
                'requestHeaders' => $requestHeaders,
                'requestBody' => $requestBody,
                'responseBody' => $responseBody
            ];

            $ch = curl_init("http://localhost:$logPort/log");
            curl_setopt($ch, CURLOPT_POST, true);
            curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($logData));
            curl_setopt($ch, CURLOPT_HTTPHEADER, ['Content-Type: application/json']);
            curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
            curl_setopt($ch, CURLOPT_TIMEOUT, 2);
            curl_exec($ch);
            curl_close($ch);
        } catch (Exception $e) {
            // Silently fail if visualizer is not running
        }
    }

    /**
     * Generic function to create a transaction.
     */
    public static function createTransaction($endpoint, $body)
    {
        $baseUrl = $_ENV['PAYLABS_BASE_URL'];
        $merchantId = $_ENV['MERCHANT_ID'];
        $privateKey = $_ENV['MERCHANT_PRIVATE_KEY'];

        [$signature, $timestamp] = self::generateSignature('POST', $endpoint, $body, $privateKey);

        $headers = [
            'Content-Type: application/json',
            'X-PARTNER-ID: ' . $merchantId,
            'X-TIMESTAMP: ' . $timestamp,
            'X-SIGNATURE: ' . $signature,
            'X-REQUEST-ID: ' . $body['requestId'],
        ];

        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $baseUrl . $endpoint);
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($body));
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);

        $response = curl_exec($ch);
        $respData = json_decode($response, true);
        
        // Log to local server
        self::logToLocalServer($endpoint, $headers, $body, $respData);
        
        if (curl_errno($ch)) {
            echo 'Error: ' . curl_error($ch) . "\n";
        } else {
            echo "Response: " . json_encode($respData, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES) . "\n";
        }

        curl_close($ch);
    }

    /**
     * Get Public IP.
     */
    public static function getPublicIp()
    {
        try {
            $context = stream_context_create(['http' => ['timeout' => 2]]);
            $json = @file_get_contents('https://api.ipify.org?format=json', false, $context);
            if ($json === false) return '127.0.0.1';
            $data = json_decode($json, true);
            return $data['ip'] ?? '127.0.0.1';
        } catch (Exception $e) {
            return '127.0.0.1';
        }
    }

    /**
     * Create SNAP Transaction.
     */
    public static function createTransactionSnap($endpoint, $body)
    {
        $baseUrl = $_ENV['PAYLABS_BASE_URL'];
        $merchantId = $_ENV['MERCHANT_ID'];
        $privateKey = $_ENV['MERCHANT_PRIVATE_KEY'];
        
        $externalId = $body['externalId'] ?? $body['requestId'] ?? self::generateRequestId();
        $ipAddress = self::getPublicIp();
        echo "Public IP: {$ipAddress}\n";

        // SNAP signature uses a modified endpoint logic
        $signatureEndpoint = $endpoint;
        if (strpos($endpoint, '/api/v1.0') === 0) {
            $signatureEndpoint = str_replace('/api/v1.0', '', $endpoint);
        }

        [$signature, $timestamp] = self::generateSignature('POST', $signatureEndpoint, $body, $privateKey);

        $headers = [
            'Content-Type: application/json;charset=utf-8',
            'X-PARTNER-ID: ' . $merchantId,
            'X-TIMESTAMP: ' . $timestamp,
            'X-SIGNATURE: ' . $signature,
            'X-EXTERNAL-ID: ' . $externalId,
            'X-IP-ADDRESS: ' . $ipAddress,
        ];

        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $baseUrl . $endpoint);
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($body));
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);

        $response = curl_exec($ch);
        $respData = json_decode($response, true);

        // Log to local server
        self::logToLocalServer($endpoint, $headers, $body, $respData);
        
        if (curl_errno($ch)) {
            echo 'Error: ' . curl_error($ch) . "\n";
        } else {
            echo "Response: " . json_encode($respData, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES) . "\n";
        }

        curl_close($ch);
    }
}
